asdf/asdf.py

```diff
@@ -1344,7 +1344,57 @@ def write_to(
             ``asdf.get_config().array_inline_threshold``.
 
         """
-
+        with config_context() as config:
+            if "all_array_storage" in kwargs:
+                config.all_array_storage = kwargs["all_array_storage"]
+            if "all_array_compression" in kwargs:
+                config.all_array_compression = kwargs["all_array_compression"]
+            if "compression_kwargs" in kwargs:
+                config.all_array_compression_kwargs = kwargs["compression_kwargs"]
+
+            used_blocks = self._blocks._find_used_blocks(self.tree, self, remove=False)
+
+            naf = AsdfFile(
+                {},
+                uri=self._uri,
+                extensions=self.extensions,
+                version=self.version,
+                ignore_version_mismatch=self._ignore_version_mismatch,
+                ignore_unrecognized_tag=self._ignore_unrecognized_tag,
+                ignore_implicit_conversion=self._ignore_implicit_conversion,
+            )
+            naf._tree = self.tree  # avoid an extra validate
+            # copy over block storage and other settings
+            block_to_key_mapping = {v: k for k, v in self._blocks._data_to_block_mapping.items()}
```

Comment:
Good question.

Within this scope, the `block_to_key_mapping` is only used for blocks that don't have associated 'data' (and instead are relying on callbacks, like with the converters). This would mean that `_data_to_block_mapping` contains only key->block associations added by `assign_block_key`.

It does look possible for a Converter to call `assign_block_key` with the same index but different keys. This would be an error in how the custom converter was written as they currently cannot support sharing blocks across objects (I don't see how `find_block_index` could allow for writing the same block with multiple keys).

I separated the _data_to_block and _key_to_block mappings in this commit and added a test for the above case:
https://github.com/braingram/asdf/commit/389792df322791c29a05a60d74682736e6046bd2